neural_modelling/src/delay_extension/delay_extension.c

```diff
@@ -393,21 +437,106 @@ static inline void spike_process(void) {
             n_packets_dropped_due_to_invalid_key += 1;
             log_debug("Invalid spike key 0x%08x", s);
         }
+        state = spin1_int_disable();
     }
 
-    // reactivate interrupts as critical section complete
+    spike_processing = false;
     spin1_mode_restore(state);
 }
 
+static void user_callback(UNUSED uint unused0, UNUSED uint unused1) {
+    spike_process();
+}
+
+static void background_callback(uint local_time, UNUSED uint timer_count) {
+    // reset the tdma for this next cycle.
+    tdma_processing_reset_phase();
+
+    // Loop through delay stages
+    for (uint32_t d = 0; d < num_delay_stages; d++) {
+        // If any neurons emit spikes after this delay stage
+        bit_field_t delay_stage_config = neuron_delay_stage_config[d];
+        if (nonempty_bit_field(delay_stage_config, neuron_bit_field_words)) {
+            // Get key mask for this delay stage and it's time slot
+            uint32_t delay_stage_delay = (d + 1) * n_delay_in_a_stage;
+            if (local_time >= delay_stage_delay) {
+                uint32_t delay_stage_time_slot =
+                        (local_time - delay_stage_delay) & num_delay_slots_mask;
+                uint8_t *delay_stage_spike_counters =
+                        spike_counters[delay_stage_time_slot];
+
+                log_debug("%u: Checking time slot %u for delay stage %u",
+                        local_time, delay_stage_time_slot, d);
+
+                // Loop through neurons
+                for (uint32_t n = 0; n < num_neurons; n++) {
+
+                    // If this neuron emits a spike after this stage
+                    if (bit_field_test(delay_stage_config, n)) {
+
+                        // Calculate key all spikes coming from this neuron will be
+                        // sent with
+                        uint32_t neuron_index = ((d * num_neurons) + n);
+                        uint32_t spike_key = neuron_index + key;
+
+                        if (delay_stage_spike_counters[n] > 0) {
```

Comment:
Is it better to put this inside a #if LOG_LEVEL >= LOG_DEBUG rather than having an if (...) that will always be checked but then only to print something using log_debug(..) ?

Reply:
Yes, definitely - done